pkg/store/labelpb/label.go

```diff
@@ -292,30 +292,37 @@ func (m *ZLabel) Compare(other ZLabel) int {
 //
 // In case of existing labels already present in given label set, it will be overwritten by external one.
 // NOTE: Labels and extend has to be sorted.
-func ExtendSortedLabels(lset, extend labels.Labels) labels.Labels {
-	ret := make(labels.Labels, 0, len(lset)+len(extend))
-
-	// Inject external labels in place.
-	for len(lset) > 0 && len(extend) > 0 {
-		d := strings.Compare(lset[0].Name, extend[0].Name)
-		if d == 0 {
-			// Duplicate, prefer external labels.
-			// NOTE(fabxc): Maybe move it to a prefixed version to still ensure uniqueness of series?
-			ret = append(ret, extend[0])
-			lset, extend = lset[1:], extend[1:]
-		} else if d < 0 {
-			ret = append(ret, lset[0])
-			lset = lset[1:]
-		} else if d > 0 {
-			ret = append(ret, extend[0])
-			extend = extend[1:]
+func ExtendSortedLabels(lset labels.Labels, extend []labels.Label) labels.Labels {
+	b := labels.NewScratchBuilder(lset.Len() + len(extend))
+	lset.Range(func(l labels.Label) {
+		for {
+			if len(extend) == 0 {
```

Comment:
Can we have a `len(extend) == 0` check outside the function?
This should help with Cortex case where there is no external labels at all. In this case no need to go through the lset loop

Reply:
Oh mb... We need to copy labels anyway so probably still need to have the check.

Reply:
I added a fast path that does `lset.Copy()` if extend is empty.